base/array.jl

```diff
@@ -1180,18 +1182,21 @@ push!(a::AbstractVector, iter...) = append!(a, iter)
 
 append!(a::AbstractVector, iter...) = foldl(append!, iter, init=a)
 
-function _append!(a, ::Union{HasLength,HasShape}, iter)
+function _append!(a::AbstractVector, ::Union{HasLength,HasShape}, iter)
```

Comment:
I wonder if we should move these generic definitions into `base/abstractarray.jl` and define simple definition of `append!` for `Vector`-inputs here (same for `prepend!`). Any thoughts?

Reply:
Special casing a concrete implementation should IMO always be preferrable over assuming more of a generic version. It's philosophically the same reason why `@inbounds` in a function accepting `AbstractArray` is bad - there's no guarantee that the concrete type isn't doing any funny business in indexing; the local information cannot be enough to ensure the use is safe.

Reply:
A potential drawback on the design is that there will be multiple potential method matches in type-unstable code for those functions, that may lead to worse inlining and optimization.

Well, we may want to accept that since the type unstable code will be slow anyway.

Reply:
I think prioritizing correctness over potentially broken fallbacks is a better bet any day of the week.